### DISCUSSION TITLE
import conventions

### DISCUSSION THREAD

**Discussion by georgebisbas** [xdslproject/xdsl#193] [Ideas]
Initiating discussion about a convention on import order

**Comment by superlopuh** [comment]
Looks like there's no way to do this automatically: https://github.com/google/yapf/issues/813

I'm happy to take this if no-one else feels like it, it's a nit but can be therapeutic to do sometimes. I also think we should do it in a commit that is ignored by git blame.

**Reply by webmiche** [reply]
There seem to be some tools that claim to do this (at least somewhat) (https://pypi.org/project/reorder-python-imports/ and https://pypi.org/project/isort/).

**Reply by webmiche** [reply]
Also https://pypi.org/project/pylint/ seems to do something like that.

**Comment by webmiche** [comment]
Let me link the [issue](https://github.com/xdslproject/xdsl/issues/147) here :)

**Comment by webmiche** [comment]
Also, I want to mention that we have some weird behavior around imports due to how mlirs python bindings work. Effectively, we have some imports that need to be order properly, which might just be bad engineering, but we did not find a better approach at the time, IIRC.

**Comment by georgebisbas** [comment]
The convention we follow in Devito is:
```
# first import specific parts of other tools
from functools import partial

# Then import tools as a whole
import numpy as np

# Then import project specific things
from devito.core.operator import CoreOperator, CustomOperator, ParTile
from devito.exceptions import InvalidOperator
```

**Comment by superlopuh** [comment]
We have this working! Closing
